walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> requiredTimeRanges = findTimeSlotsForNewMeeting(
+        filterRelevantEvents(events, request.getAttendees()),
+        request.getDuration());
+    List<TimeRange> optionalTimeRanges = findTimeSlotsForNewMeeting(
+        filterRelevantEvents(events, request.getOptionalAttendees()),
+        request.getDuration());
+    List<TimeRange> intersection = findTimeRangesIntersections(
+        requiredTimeRanges, optionalTimeRanges, request.getDuration());
+    if (intersection.isEmpty()) {
+      return requiredTimeRanges;
+    }
+    return intersection;
+  }
+
+  private static List<TimeRange> findTimeRangesIntersections(List<TimeRange> required,
+                                                             List<TimeRange> optional,
+                                                             long duration) {
+    List<TimeRange> intersection = new ArrayList<>();
+    for (TimeRange requiredTimeRange : required) {
+      for (TimeRange optionalTimeRange : optional) {
+        createIntersection(requiredTimeRange, optionalTimeRange)
+            .filter(timeRange -> timeRange.duration() >= duration)
+            .ifPresent(intersection::add);
+      }
+    }
+    return intersection;
+  }
+
+  private static Optional<TimeRange> createIntersection(TimeRange a, TimeRange b) {
+    if (a.overlaps(b)) {
+      return Optional.of(TimeRange.fromStartEnd(
+          Math.max(a.start(), b.start()),
+          Math.min(a.end(), b.end()),
+          false));
+    }
+    return Optional.empty();
+  }
+
+  private static List<TimeRange> findTimeSlotsForNewMeeting(Collection<Event> events,
+                                                            long duration) {
+    List<Event> sortedEvents = sortEvents(events);
+    List<TimeRange> queryResult = new ArrayList<>();
+    int currentTime = 0;
+    for (Event event : sortedEvents) {
+      if (currentTime + duration <= event.getWhen().start()) {
+        int end = (int) Math.max(event.getWhen().start(), currentTime + duration);
+        queryResult.add(TimeRange.fromStartEnd(currentTime, end, false));
+      }
+      currentTime = Math.max(currentTime, event.getWhen().end());
+    }
+
+    if (currentTime + duration <= TimeRange.END_OF_DAY) {
+      queryResult.add(TimeRange.fromStartEnd(currentTime, TimeRange.END_OF_DAY, true));
+    }
+
+    return queryResult;
+  }
+
+  private static List<Event> filterRelevantEvents(Collection<Event> events,
+                                                  Collection<String> requestAttendees) {
+    return events.stream()
+        .filter(event -> event.getAttendees().stream().anyMatch(requestAttendees::contains))
+        .collect(Collectors.toList());
+  }
+
+  private static List<Event> sortEvents(Collection<Event> events) {
```

Comment:
I've used Collections.sort in previous commits and changed it to sorted() because of two reasons:
- It works only on List not every Collection (in this case it's not a problem, because the argument applied is a List, but it may lead to bugs if I assume that)

- **(main reason)** It modifies given as argument list. I've [read](https://stackoverflow.com/questions/12019978/is-it-a-good-practice-to-change-arguments-in-java) that modyfing arguments is not good, so this is why I changed it.
Could you please explain why would it be better to use Collections.sort in this case?

Reply:
All valid concerns!
It's a bit of a philosophical question here: the collection we pass to `sortEvents` is basically in all _current_ cases a list (created by `filterRelevantEvents`), so one could argue that we don't change arguments. If we would change the visibility of `sortEvents` to `public`, like in a later PR, the story would be different.

We can leave it as it is, great thought process :-)